### PR TITLE
Implementing `karmadactl promote` to promote workloads from legacy clusters to karmada

### PR DESCRIPTION
**Pull request by lonelyCZ**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
part of #1039 

**Special notes for your reviewer**:
first stage of https://github.com/karmada-io/karmada/discussions/1039#discussioncomment-1726426

```bash
[root@master67 karmada]# karmadactl promote -h
Promote resources from legacy clusters to karmada control plane.

Usage:
  karmadactl promote <RESOURCE_TYPE> <RESOURCE_NAME> -n <NAME_SPACE> -c <CLUSTER_NAME> [flags]

Examples:

# Promoting deployment(default/nginx) from cluster1 to karmada.
karmadactl promote deployment nginx -n default -c cluster1

# Promoting deployment((default/nginx) with gvk from cluster1 to karmada.
karmadactl promote deployment.v1.apps nginx -n default -c cluster1

# Only printing the converted resource, not to create them in karmada.
karmadactl promote deployment nginx -n default -c cluster1 -o yaml|json

# Promoting secret(default/default-token) from cluster1 to karmada.
karmadactl promote secret default-token -n default -c cluster1

# If it's cluster of pull mode, you can use --cluster-kubeconfig and --cluster-context
karmadactl promote deployment nginx -n default -c cluster1 --cluster-kubeconfig=<CLUSTER_KUBECONFIG_PATH> --cluster-context=<CLUSTER_CONTEXT>
```

Printing converted resources.
```bash
[root@master67 karmada]# karmadactl promote deploy nginx -c bj-47 -o yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"labels":{"app":"nginx"},"name":"nginx","namespace":"default"},"spec":{"replicas":1,"selector":{"matchLabels":{"app":"nginx"}},"template":{"metadata":{"labels":{"app":"nginx"}},"spec":{"containers":[{"image":"nginx","name":"nginx"}]}}}}
    work.karmada.io/conflict-resolution: overwrite
  labels:
    app: nginx
  name: nginx
  namespace: default
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: nginx
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx
        imagePullPolicy: Always
        name: nginx
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
---
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  creationTimestamp: null
  name: nginx-propagation
  namespace: default
spec:
  placement:
    clusterAffinity:
      clusterNames:
      - bj-47
  resourceSelectors:
  - apiVersion: apps/v1
    kind: Deployment
    name: nginx
```

**Does this PR introduce a user-facing change?**:

```release-note
karmadactl promote <RESOURCE_TYPE> <RESOURCE_NAME> -n <NAME_SPACE> -c <CLUSTER_NAME> [flags]
```
/cc @RainbowMango @GitHubxsy @mrlihanbo @pigletfly @iawia002 @dddddai

### CODE REVIEW

**Review by GitHubxsy — APPROVED** [review]
/lgtm

File: pkg/karmadactl/promote.go [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+	karmadaclientset "github.com/karmada-io/karmada/pkg/generated/clientset/versioned"
+	"github.com/karmada-io/karmada/pkg/karmadactl/options"
+	"github.com/karmada-io/karmada/pkg/resourceinterpreter/defaultinterpreter/prune"
+	"github.com/karmada-io/karmada/pkg/util/gclient"
+	"github.com/karmada-io/karmada/pkg/util/restmapper"
+)
+
+var (
+	promoteShort = `Promote resources from legacy clusters to karmada control plane.`
+	promoteLong  = `Promote resources from legacy clusters to karmada control plane.`
```
Comment (RainbowMango) [inline pkg/karmadactl/promote.go:37]:
```suggestion
	promoteLong  = `Promote resources from legacy clusters to karmada control plane. Requires the cluster be joined or registered.`
```

File: pkg/karmadactl/promote.go [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+
+	flag := cmd.Flags()
+	opts.AddFlags(flag)
+
+	return cmd
+}
+
+func getPromoteExample(parentCommand string) string {
+	promoteExample := `
+# Promoting deployment(default/nginx) from cluster1 to karmada.
```
Comment (RainbowMango) [inline pkg/karmadactl/promote.go:73]:
```suggestion
# Promote deployment(default/nginx) from cluster1 to karmada.
```
Same as below.

File: pkg/karmadactl/promote.go [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+` + parentCommand + ` promote deployment nginx -n default -c cluster1
+
+# Promoting deployment((default/nginx) with gvk from cluster1 to karmada.
+` + parentCommand + ` promote deployment.v1.apps nginx -n default -c cluster1
+
+# Only printing the converted resource, not to create them in karmada.
```
Comment (RainbowMango) [inline pkg/karmadactl/promote.go:79]:
```suggestion
# Dumps the artifacts but does not deploy them to Karmada, same as 'dry run'.
```

File: pkg/karmadactl/promote.go [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+` + parentCommand + ` promote deployment nginx -n default -c cluster1 -o yaml|json
+
+# Promoting secret(default/default-token) from cluster1 to karmada.
+` + parentCommand + ` promote secret default-token -n default -c cluster1
+
+# If it's cluster of pull mode, you can use --cluster-kubeconfig and --cluster-context
```
Comment (RainbowMango) [inline pkg/karmadactl/promote.go:85]:
```suggestion
# For clusters with 'Pull' mode, use '--cluster-kubeconfig' and `--cluster-context` to specify the configuration.
```

Comment (RainbowMango) [reply pkg/karmadactl/promote.go:85]: I think we can add one more example for pull mode clusters, as the `--cluster-context` could be omitted.

File: pkg/karmadactl/promote.go (hunk 3 above)
Comment (RainbowMango) [reply pkg/karmadactl/promote.go:79]: Also, replace `karmada` to `Karmada`.

File: pkg/karmadactl/promote.go [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+		_, err = karmadaClient.PolicyV1alpha1().PropagationPolicies(opts.Namespace).Create(context.TODO(), pp, metav1.CreateOptions{})
+
+		return err
+	}
+	if err != nil {
+		return fmt.Errorf("failed to get PropagationPolicy(%s/%s) in control plane: %v", opts.Namespace, name, err)
+	}
+
+	// PropagationPolicy already exists, not to create it
+	// TODO: support to auto edit PropagationPolicy if it already exists
```
Comment (RainbowMango) [inline pkg/karmadactl/promote.go:402]: I don't suppose we gonna support that dangerous operation.

Comment (RainbowMango) [reply pkg/karmadactl/promote.go:402]: Same as the ClusterPropagationPolicy.

**Review by RainbowMango — COMMENTED** [review]
Looks pretty good, just some nits from me.

File: pkg/karmadactl/promote.go (hunk 5 above)
Comment (lonelyCZ) [reply pkg/karmadactl/promote.go:402]: Thanks for your reviews. I will do it.

File: pkg/karmadactl/promote.go (hunk 4 above)
Comment (mrlihanbo) [reply pkg/karmadactl/promote.go:85]: how about to add example of --dry-run

Comment (RainbowMango) [reply pkg/karmadactl/promote.go:85]: I don't know if it's too complicated.

Do you want `--dry-run` to dump the artifests?

Comment (lonelyCZ) [reply pkg/karmadactl/promote.go:85]: `--dry-run` is a general concept in `kubectl` and `karmadactl` that seemly need not to be a example.

**Review by RainbowMango — COMMENTED** [review]
/approve

**Pull request merged**